core/src/main/scala/tamer/kafka/Kafka.scala

```diff
@@ -26,131 +27,179 @@ object StateKey {
+      log: LogWriter[Task]
+  ): RIO[R with Blocking with Clock, Unit] =
+    dataQueue
+      .map { case (k, v) => new ProducerRecord(sinkTopic, k, v) }
+      .mapChunksM(recordChunk =>
+        valueProducerService
+          .produceChunkAsync(recordChunk)
+          .retry(tenTimes)
+          .flatten <* log.info(s"pushed ${recordChunk.size} messages to $sinkTopic")
+      )
+      .runDrain
+      .onError(e => log.warn(s"Could not push data to topic '$sinkTopic': ${e.prettyPrint}").orDie)
+
+  case class Live[K, V, S](
+      config: Config.Kafka,
+      setup: SourceConfiguration[K, V, S],
+      stateTransitionFunction: (S, Queue[Chunk[(K, V)]]) => ZIO[Any, TamerError, S],
+      stateConsumerService: Consumer.Service,
+      stateProducerService: Producer.Service[Registry with Topic, StateKey, S],
+      valueProducerService: Producer.Service[Registry with Topic, K, V]
+  ) extends Service {
+    private[this] final val logTask: Task[LogWriter[Task]] = log4sFromName.provide("tamer.kafka")
+
+    private final def mkRegistry(src: SchemaRegistryClient, topic: String): ZLayer[Any, Nothing, Registry with Topic] =
+      (ZLayer.succeed(src) >>> Registry.live) ++ (ZLayer.succeed(topic) >>> Topic.live)
+
+    final def source(
+        dataQueue: Queue[Chunk[(K, V)]],
+        layer: ULayer[Registry with Topic],
+        log: LogWriter[Task]
+    ): ZStream[Blocking with Clock, Throwable, Offset] = {
+      sealed trait ExistingState
+      case object PreexistingState extends ExistingState
+      case object EmptyState       extends ExistingState
+      val stateKey                                                  = StateKey(setup.tamerStateKafkaRecordKey.toHexString, config.state.groupId)
+      def waitNonemptyAssignment(consumerService: Consumer.Service) = consumerService.assignment.withFilter(_.nonEmpty).retry(tenTimes)
+      val stateTopicSub                                             = Subscription.topics(config.state.topic)
+      def mkRecord(k: StateKey, v: S)                               = new ProducerRecord(config.state.topic, k, v)
+      def containsExistingState(partitionSet: Set[TopicPartition]) =
+        stateConsumerService.endOffsets(partitionSet).map(_.values.exists(_ > 0L)).map(if (_) PreexistingState else EmptyState)
+      val subscribeToExistingState =
+        stateConsumerService.subscribe(stateTopicSub) *> waitNonemptyAssignment(stateConsumerService) >>= containsExistingState
+      val stateKeySerde = Serde[StateKey](isKey = true)(AvroCodec.codec[StateKey])
+      val initializeAssignedPartitions: ZIO[Blocking with Clock, Throwable, Unit] = subscribeToExistingState.flatMap {
+        case PreexistingState => log.info(s"consumer group ${config.state.groupId} resuming consumption from ${config.state.topic}")
+        case EmptyState =>
+          log.info(
+            s"consumer group ${config.state.groupId} never consumed from ${config.state.topic}, setting offset to earliest"
```

Comment:
if my comment is correct this could become "never consumed from $topic:$partition"